pkg/gardenlet/operation/botanist/dualstackmigration.go

```diff
@@ -0,0 +1,118 @@
+// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package botanist
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+)
+
+func (b *Botanist) checkInfraStatus(ctx context.Context) (bool, error) {
+	infra, err := b.Shoot.Components.Extensions.Infrastructure.Get(ctx)
+	if err != nil {
+		return false, fmt.Errorf("failed to get infra resource: %w", err)
+	}
+	result := len(infra.Status.Networking.Nodes) == 2
+	return result, nil
+}
+
+func (b *Botanist) checkNetworkStatusIPFamilies(ctx context.Context) (bool, error) {
+	network, err := b.Shoot.Components.Extensions.Network.Get(ctx)
+	if err != nil {
+		return false, fmt.Errorf("failed to get network resource: %w", err)
+	}
+	providerStatus := network.Status.GetProviderStatus()
+	if providerStatus.Raw == nil {
+		return false, fmt.Errorf("network providerStatus is nil")
+	}
+	var networkStatus map[string]any
+	if err := json.Unmarshal(providerStatus.Raw, &networkStatus); err != nil {
+		return false, fmt.Errorf("failed to unmarshal network providerStatus: %w", err)
+	}
+	ipFamilies, ok := networkStatus["ipFamilies"]
```

Comment:
This doesn't look good. You are making Gardener dependent on a field that is controlled by `Network` extensions. If this is a necessary contract, you should rather add `IPFamilies` as a regular field to the `extensions.gardener.cloud/v1alpha1.NetworkStatus` and read it from there. Unmarshaling the `providerConfig` or `providerStatus` in `gardener/gardener` is violating the extension concepts.

Reply:
Yes, changed it to use `NetworkStatus`.